### ISSUE TITLE
[feature] Gitlab merged results and merge train support

### ISSUE THREAD

**Issue by matray** [moonrepo/moon#1412]
**Is your feature request related to a problem? Please describe.**

When invoking moon using a GitLab [merged_results](https://docs.gitlab.com/ee/ci/pipelines/merged_results_pipelines.html) or [merge_train](https://docs.gitlab.com/ee/ci/pipelines/merge_trains.html) pipeline, moon fails to execute due to the unique ref nature of these "shadow commits".

```text
Error: process::failed
  × Process git failed with a 128 exit code. 
  │ 
  │ fatal: ambiguous argument 'main': unknown revision or path not in the
  │ working tree.
  │ Use '--' to separate paths from revisions, like this:
  │ 'git <command> [<revision>...] -- [<file>...]'
```
**Describe the solution you'd like**

Ideally, when moon is configured for GitLab as the provider, it uses the following configuration as precedence

MOON_BASE: $CI_MERGE_REQUEST_DIFF_BASE_SHA
MOON_HEAD: $CI_MERGE_REQUEST_SOURCE_BRANCH_SHA
    
The existence of these env vars is predicated upon 

```text
 $CI_MERGE_REQUEST_EVENT_TYPE == "merged_result" || $CI_MERGE_REQUEST_EVENT_TYPE == "merge_train"
```

Gitlab has a solid reference for these variables that can be found in environment here: https://docs.gitlab.com/ee/ci/variables/predefined_variables.html

**Describe alternatives you've considered**

Setting these manually or via the CLI seems to work as a workaround.

**Additional context**

Nothing specific other than using an obscure-ish pipeline concept specific to gitlab

**Comment by milesj** [comment]
Started on this here: https://github.com/milesj/rust-cicd-env/pull/5

Will be in the next version.

**Comment by milesj** [comment]
Let's go https://github.com/moonrepo/moon/pull/1430

**Comment by milesj** [comment]
Landed in 1.24
